crates/docs/src/validation.rs

```diff
@@ -90,26 +142,75 @@ pub fn get_parent_dir(levels_up: usize) -> PathBuf {
     dir
 }
 
-pub fn assert_valid_snippet(
-    condition: bool,
-    snippet: &Snippet,
-    tool_name: &str,
-    err_message: &str,
-) {
+pub fn assert_valid_snippet(condition: bool, snippet: &Snippet, err_message: &str) {
     assert!(
         condition,
-        "Found invalid {} snippet in the docs in file: {} at line {}\n{}",
-        tool_name, snippet.file_path, snippet.line_start, err_message
+        "Found invalid {} snippet in the docs in at {}:{}:1\n{}",
+        snippet.snippet_type.as_str(),
+        snippet.file_path,
+        snippet.line_start,
+        err_message
     );
 }
 
 pub fn print_success_message(snippets_len: usize, tool_name: &str) {
     println!("Successfully validated {snippets_len} {tool_name} docs snippets");
 }
 
-pub fn print_skipped_snippet_message(snippet: &Snippet, tool_name: &str) {
+pub fn print_skipped_snippet_message(snippet: &Snippet) {
     println!(
         "Skipped validation of {} snippet in the docs in file: {} at line {}",
-        tool_name, snippet.file_path, snippet.line_start
+        snippet.snippet_type.as_str(),
+        snippet.file_path,
+        snippet.line_start,
     );
 }
+
+pub fn create_listings_to_packages_mapping() -> HashMap<String, Vec<String>> {
+    let docs_listings_path = "../../docs/listings";
+    let mut mapping = HashMap::new();
+
+    for listing in WalkDir::new(docs_listings_path)
+        .min_depth(1)
+        .max_depth(1)
+        .into_iter()
+        .filter_map(Result::ok)
+        .filter(|e| e.file_type().is_dir())
+    {
+        let listing_path = listing.path();
+        let crates_dir = listing_path.join("crates");
+
+        if crates_dir.is_dir() {
+            let packages = list_packages_in_directory(&crates_dir);
+            if let Some(listing_name) = listing_path.file_name().and_then(|x| x.to_str()) {
+                if !packages.is_empty() {
+                    mapping.insert(listing_name.to_string(), packages);
+                }
+            }
+        }
+    }
+
+    mapping
+}
+
+fn list_packages_in_directory(dir_path: &Path) -> Vec<String> {
+    let crates_path = dir_path.join("");
```

Comment:
```suggestion
    let crates_path = dir_path.to_owned();
```

Also why you need this at all?

Reply:
It's used to retrieve packages located in docs listing directory.